[LEP] Asynchronous pull backup volumes into cluster CR (#2623)
longhorn/longhorn: enhancements/20210525-async-pull-backups.md

The current [longhorn/backupstore](https://github.com/longhorn/backupstore) list and inspect command behavior are:
- `backup ls --volume-only`: List all backup volumes and read it's metadata (`volume.cfg`).
- `backup ls --volume <volume-name>`: List all historical volume backups and read it's metadata (`backup_backup_<backup-hash>.cfg`).
- `backup ls --volume <volume-name> --volume-only`: Read a single backup volume metadata (`volume.cfg`).
- `backup inspect <backup>`: Read a single volume's backup metadata (`backup_backup_<backup-hash>.cfg`).

After this enhancement, the [longhorn/backupstore](https://github.com/longhorn/backupstore) list and inspect command behavior are:
- `backup ls --volume-only`: List all backup volume names.
- `backup ls --volume <volume-name>`: List all historical volume backup names.
- `backup ls --volume <volume-name> --volume-only`: Read a single backup volume name.
- `backup inspect <metadata-path>`: Read a single backup volume metadata (`volume.cfg`) or read a single volume's backup metadata (`backup_backup_<backup-hash>.cfg`).

Generally speaking, we want to separate the **list** and **read** command.

## Design

### Implementation Overview

> We'll create a new Custom Resource Definition (CRD) called _backups`_ to save pulled backup volumes and its historical volume backups as the Custom Resource (CR).

Reviewer: Can you talk more about this CRD?
Will Longhorn create one CR for each backup or backup volume record in the external backupstore? After reading this LEP, I still have no basic picture about this new CRD

Author: I've updated it.

> There is already a backup store monitor that runs as a timer, the timer period is the setting `backupstore-poll-interval`. Within it, it runs:

Reviewer: 1. Need to consider the modification of the backup target setting. If the setting is changed, all CRs should be removed, right?
2. Need to think about the CR update conflict, and how Longhorn balances the below workload to all workers of all nodes. Do we need a new controller for the new CRD?

Author: > 1. Need to consider the modification of the backup target setting. If the setting is changed, all CRs should be removed, right?

I'd prefer to remove it when running synchronous in the controller, that is:
- Backup Volume:
1. Pull the backup volumes that are in the backup store and aren't in the cluster CR, create a new CR.
2. Delete the backup volumes that are in the cluster CR and aren't in the backup store, delete the CR.
- Volume Backups:
1. Calculate the difference set of the volume backups that are in the backup store and aren't in the cluster CR, add it to the backup volume CR's `spec.backups`.
2. Calculate the difference set of the volume backups that are in the cluster CR and aren't in the backup store, delete it from the backup volume CR's `spec.backups`
3. Update the backup volume CR

> 2. Need to think about the CR update conflict, and how Longhorn balances the below workload to all workers of all nodes. Do we need a new controller for the new CRD? I think Joshua mentions some ideas in a Github ticket. You can refer to it first.

Now, the backup store monitor running as a goroutine in the setting controller. My current thought is to keep it as-is.
I'm fine to create a new controller _or_ keep it in the current setting controller. WDYT?

I don't fully understand the question `how Longhorn balances the below workload to all workers of all nodes`. But I'll try to answer my understanding. The current [code](https://github.com/longhorn/longhorn-manager/blob/254aa17/controller/setting_controller.go#L645-L677) would run the workload randomly, this is how Joshua enhance it at v1.1.1 to prevent each node access the external backup store.

- `backup ls --volume <volume-name> --volume-only`: Read a single backup volume metadata (`volume.cfg`).
- `backup inspect <backup>`: Read a single volume's backup metadata (`backup_backup_<backup-hash>.cfg`).

After this enhancement, the [longhorn/backupstore](https://github.com/longhorn/backupstore) list and inspect command behavior are:
- `backup ls --volume-only`: List all backup volume names.
- `backup ls --volume <volume-name>`: List all historical volume backup names.
- `backup ls --volume <volume-name> --volume-only`: Read a single backup volume name.
- `backup inspect <metadata-path>`: Read a single backup volume metadata (`volume.cfg`) or read a single volume's backup metadata (`backup_backup_<backup-hash>.cfg`).

Generally speaking, we want to separate the **list** and **read** command.

## Design

### Implementation Overview

We'll create a new Custom Resource Definition (CRD) called `backupvolumes.longhorn.io` to save pulled backup volumes metadata 
and its historical volume backups metadata as the Custom Resource (CR). For each backup volume, we'll create a new CR.
- `metadata.name`: the backup volume name.

> - `status.name`: the name field inside the backup volume metadata.

Reviewer: Remove this one. The backup volume name is already in `metadata`

   6. Loop the `volumeBackupsToPull`, read the historical volume backup metadata from the external backup store, and add to the backup volume CR field `status.backups`.
   7. Loop the `volumeBackupsToDelete`, delete the backup volume CR `status.backups`.
   8. Update the backup volume (BV) CR `status.backups`.

For the longhorn manager endpoints:
- **GET** `/v1/backupvolumes`: read all backup volumes from the cluster custom resource (CR).
- **GET** `/v1/backupvolumes/{volName}`: read a single backup volume from the cluster custom resource (CR).
- **DELETE** `/v1/backupvolumes/{volName}`: delete a single backup volume from the cluster custom resource (CR) and also from external backup store.
- **GET** `/v1/backupvolumes/{volName}?action=backupList`: read a single volume backups from the cluster custom resource (CR).
- **GET** `/v1/backupvolumes/{volName}?action=backupGet`: read a single volume backup from the cluster custom resource (CR).
- **DELETE** `/v1/backupvolumes/{volName}?action=backupDelete`: delete a single volume backup from the cluster custom resource (CR) and also from external backup store.

### Test plan

> 

Reviewer: Test the case: 2 clusters share the same backupstore

Author: Good suggestion, thank you.

3. When the user deletes a backup volume on the Longhorn GUI:
   1. list backup volumes on the Longhorn GUI, the deleted one does not exist immediately
   2. check the external backup store, the backup volume will be deleted after a while.
4. When the user deletes a historical volume backup on the Longhorn GUI:
   1. list historical volume backups on the Longhorn GUI, the deleted one does not exist immediately
   2. check the external backup store, the historical backup will be deleted after a while.
5. When the user deletes a backup volume on the external backup store manually.
   After `backupstore-poll-interval` seconds, list backup volumes on the Longhorn GUI, the deleted one does not exist.
6. When the user deletes a historical volume backup on the external backup store manually.
   After `backupstore-poll-interval` seconds, list historical volume backups on the Longhorn GUI, the deleted one does not exist immediately.

### Upgrade strategy

Before this enhancement, the user might set `backupstore-poll-interval` to `0` to reduce the request to the external backup store.
This is majorly for the user who wants to reduce the cost on access AWS S3.

> After this enhancement, if the user ever configured the setting `backupstore-poll-interval` to `0`, then the user is unable to list backup volumes

Reviewer: Can we just add a warning in UI rather than directly forbidding listing (historical) backup volumes?

- `backup ls --volume-only`: List all backup volumes and read it's metadata (`volume.cfg`).
- `backup ls --volume <volume-name>`: List all historical volume backups and read it's metadata (`backup_backup_<backup-hash>.cfg`).
- `backup ls --volume <volume-name> --volume-only`: Read a single backup volume metadata (`volume.cfg`).
- `backup inspect <backup>`: Read a single volume's backup metadata (`backup_backup_<backup-hash>.cfg`).

After this enhancement, the [longhorn/backupstore](https://github.com/longhorn/backupstore) list and inspect command behavior are:
- `backup ls --volume-only`: List all backup volume names.
- `backup ls --volume <volume-name>`: List all historical volume backup names.
- `backup ls --volume <volume-name> --volume-only`: Read a single backup volume name.
- `backup inspect <metadata-path>`: Read a single backup volume metadata (`volume.cfg`) or read a single volume's backup metadata (`backup_backup_<backup-hash>.cfg`).

Generally speaking, we want to separate the **list** and **read** command.

## Design

### Implementation Overview

We'll create a new Custom Resource Definition (CRD) called `backups.longhorn.io` to save pulled backup volumes metadata 

> and its historical volume backups metadata as the Custom Resource (CR). For each backup volume, we'll create a new CR.

Author: Good suggestion.

- `spec.lastBackupAt`: the lastBackupAt field inside the backup volume metadata.
- `spec.dataStored`: the dataStored field inside the backup volume metadata.
- `spec.messages`: the messages field inside the backup volume metadata.
- `spec.backingImageName`: the backingImageName field inside the backup volume metadata.
- `spec.backingImageURL`: the backingImageURL field inside the backup volume metadata.
- `spec.baseImage`: the baseImage field inside the backup volume metadata.
- `spec.backups`: a map of volume backups (key is volume backup name, value is the volume backup metadata).
  - `spec.backups[volumeBackupName].name`: the name field inside a historical volume backup metadata.
  - `spec.backups[volumeBackupName].url`: the url field inside a historical volume backup metadata.
  - `spec.backups[volumeBackupName].snapshotName`: the snapshotName field inside a historical volume backup metadata.
  - `spec.backups[volumeBackupName].snapshotCreated`: the snapshotCreated field inside a historical volume backup metadata.
  - `spec.backups[volumeBackupName].created`: the created field inside a historical volume backup metadata.

>   - `spec.backups[volumeBackupName].size`: the size field inside a historical volume backup metadata.

Author: Yes, the volume backup data size.

  - `spec.backups[volumeBackupName].labels`: the labels field inside a historical volume backup metadata.
  - `spec.backups[volumeBackupName].volumeName`: the volumeName field inside a historical volume backup metadata.
  - `spec.backups[volumeBackupName].volumeSize`: the volumeSize field inside a historical volume backup metadata.
  - `spec.backups[volumeBackupName].volumeCreated`: the volumeCreated field inside a historical volume backup metadata.
  - `spec.backups[volumeBackupName].volumeBackingImageName`: the volumeBackingImageName field inside a historical volume backup metadata.

>   - `spec.backups[volumeBackupName].volumeBackingImageURL`: the volumeBackingImageURL field inside a historical volume backup metadata.

Author: Okay, I removed all backing image related fields.

>   - `spec.backups[volumeBackupName].volumeBackingImageURL`: the volumeBackingImageURL field inside a historical volume backup metadata.

Reviewer: Please retain `volumeBackingImageName`. This field is still available

# Title

> Asynchronous pull backup volumes and historical volume backups into cluster custom resource (CR).

Reviewer 2: ```suggestion
Final consistent backup volumes with historical volume backups from the remote backup store
```

## Summary

> Asynchronous way to pull backup volumes and historical volume backups from the external backup store (S3/NFS) into cluster custom resource (CR).
> This enhancement is to fix the timeout issue when listing backup volumes from the external backup store _or_ listing historical volume backups 
> from the external backup store.

Reviewer 2: ```suggestion
Currently, Longhorn uses a blocking way for communication with the remote backup store, so there will be some potential voluntary or involuntary factors (ex:  network latency) impacting the functions relying on remote backup stores like listing volume backups or even causing further [cascading problems](###related-issues) after the backup store operation.

This enhancement is to propose an asynchronous way to pull backup volumes and historical volume backups from the external backup store (S3/NFS) then persistently saved via cluster custom resources. 

This can resolve the problems above mentioned by asynchronously querying the list of backup volumes and historical volume backups from the remote backup store for final consistent available results. It's also scalable for the costly resources created by the original blocking query operations.
```

### Related Issues

- https://github.com/longhorn/longhorn/issues/1761
- https://github.com/longhorn/longhorn/issues/1955
- https://github.com/longhorn/longhorn/issues/2536
- https://github.com/longhorn/longhorn/issues/2543

## Motivation

### Goals

> Decrease the time when the user clicks list backup _or_ list historical volume backups under the circumstances that lots of volume backups _or_ 
> lots of historical volume backups in the external backup store **and** the latency from Longhorn cluster to the external backup store is high.
> 

Reviewer 2: ```suggestion
Decrease the query latency when listing volume backups _or_ historical volume backups in the circumstances like lots of volume backups, lots of historical volume backups, or the network latency between the Longhorn cluster and the remote backup store.
```

### Non-goals [optional]

Automatically adjust the external backup store pull period.

## Proposal

Currently, we have a setting `backupstore-poll-interval` to periodically pull the external backup store lastest volume backups within a controller.
We want to leverage it to pull the external backup volumes and historical volume backups from the external backup store and saves them into the cluster 

> custom resource (CR). Therefore, we'll create a new Custom Resource Definition (CRD) called _backups_ to save the backup volumes and historical volume backups.

Reviewer 2: ```suggestion
custom resource (CR). Therefore, we'll create a new Custom Resource Definition (CRD) called _backupvolumes_ to save the backup volumes and historical volume backups.
```

With this enhancement, we want to:

> 1. Pull the backup volumess and historical volume backups that **are in** the external backup store and **aren't in** the cluster CR only.

Reviewer 2: ```suggestion
1. Pull the backup volumes and historical volume backups that **are in** the external backup store and **aren't in** the cluster CR only.
```

### User Stories

Before this enhancement, when the user's environment under the circumstances that the external backup store has lots of backup volumes _or_ 
historical volume backups, and the latency between the longhorn manager to the external backup store is high.
Then if the user clicks the `Backup` on the GUI, the user might hit list backup volumes _or_ list historical volume backups timeout issue 
(the default timeout is 1 min).

We choose to not create a new setting for the user to increase the list timeout value is because the browser has its timeout value also.
Let's say the list backups needs 5 minutes to finish. Even we allow the user to increase the longhorn manager list timeout value, we can't change 
the browser default timeout value. Furthermore, some browser doesn't allow the user to change default timeout value like Google Chrome.

After this enhancement, when the user's environment under the circumstances that the external backup store has lots of backup volumes _or_ 
historical volume backups, and the latency between the longhorn manager to the external backup store is high.

> Then if the user clicks the `Backup` on the GUI, the user can list backup volumes _or_ list historical volume backups without timeout issue.

Reviewer 2: ```suggestion
Then if the user clicks the `Backup` on the GUI, the user can eventually list backup volumes _or_ list historical volume backups without timeout issue.
```

  {
    "pvc-004d8edb-3a8c-4596-a659-3d00122d3f07": {
      "Backups": {
        "s3://backupbucket@minio/?backup=backup-02224cb26b794e73\u0026volume=pvc-004d8edb-3a8c-4596-a659-3d00122d3f07": {},
        ...
        "s3://backupbucket@minio/?backup=backup-fa78d89827664840\u0026volume=pvc-004d8edb-3a8c-4596-a659-3d00122d3f07": {}
      }
    }
  }
  ```
- `backup inspect <metadata-path>`: Read a single backup volume metadata (`volume.cfg`) or read a single volume's backup metadata (`backup_backup_<backup-hash>.cfg`).
  - Read a single backup volume metdata (`volume.cfg`). For example:
    ```shell
    $ backup inspect s3://backupbucket@minio/?volume=pvc-004d8edb-3a8c-4596-a659-3d00122d3f07
    {
      "Name": "pvc-004d8edb-3a8c-4596-a659-3d00122d3f07",
      "Size": "2147483648",
      "Labels": {},
      "Created": "2021-05-12T00:52:01Z",
      "LastBackupName": "backup-c5f548b7e86b4b56",
      "LastBackupAt": "2021-05-17T05:31:01Z",
      "DataStored": "121634816",
      "Messages": {}
    }
    ```
  - Read a single volume's backup metadata (`backup_backup_<backup-hash>.cfg`). For example:
    ```shell

>     $ backup inspect s3://backupbucket@minio/?backup=backup-fa78d89827664840\u0026volume=pvc-004d8edb-3a8c-4596-a659-3d00122d3f07

Author: How about this? Since the URL and volume name is mandatory, but the backup name is optional.
```
backup inspect <URL> <volume-name> --backup <backup-name>
```

>     $ backup inspect s3://backupbucket@minio/?backup=backup-fa78d89827664840\u0026volume=pvc-004d8edb-3a8c-4596-a659-3d00122d3f07

Reviewer 2: Good to me. Actually, the mandatory option is better to use `argument` :+1:

4. When the user deletes a historical volume backup on the Longhorn GUI:
   1. list historical volume backups on the Longhorn GUI, the deleted one does not exist immediately
   2. check the external backup store, the historical backup will be deleted after a while.
5. When the user deletes a backup volume on the external backup store manually.
   After `backupstore-poll-interval` seconds, list backup volumes on the Longhorn GUI, the deleted one does not exist.
6. When the user deletes a historical volume backup on the external backup store manually.
   After `backupstore-poll-interval` seconds, list historical volume backups on the Longhorn GUI, the deleted one does not exist immediately.

### Upgrade strategy

Before this enhancement, the user might set `backupstore-poll-interval` to `0` to reduce the request to the external backup store.
This is majorly for the user who wants to reduce the cost on access AWS S3.

After this enhancement, if the user ever configured the setting `backupstore-poll-interval` to `0`, then the user is unable to list backup volumes
_or_ historical volume backups. We should add this note to the documentation and the release note.

## Note

> With this enhancement, the CR _backups.longhorn.io_ is updated when the asynchronous timer be triggered.

Reviewer 2: ```suggestion
With this enhancement, the CR _backupvolumes.longhorn.io_ is updated when the asynchronous timer be triggered.
```

     - `status.backups[volumeBackupName].labels`: the volume backup labels.
     - `status.backups[volumeBackupName].volumeName`: the backup volume name.
     - `status.backups[volumeBackupName].volumeSize`: the backup volume size.
     - `status.backups[volumeBackupName].volumeCreateTimestamp`: the backup volume creation timpstamp.
     - `status.backups[volumeBackupName].messages`: the error messages when list volume backup or inspect volume backup metadata.
   - `status.lastSyncedTimestamp`: records the last time the backup store contents were synced into the cluster.

2. At the existed `setting_controller`.
   
   We'll remove the existed code on watches the change of CR `settings.longorn.io` field `backup-target`, and `backup-target-credential-secret`, and `backupstore-poll-interval`. Also, we'll remove the backup store monitor in the setting controller. The backup store monitor run as a goroutine to 
   periodically pull the latest volume backup name and volume backup timestamp, and updates the `status.LastBackup` and `status.LastBackupAt` to the 
   CR `volumes.longhorn.io` if it's a DR volume.
   
3. Create a new controller called `backup_volume_controller`.
   

>    It watches the change of the settings `backup-target`, and `backup-target-credential-secret`, and `backupstore-poll-interval`. Within the controller, it runs:

Reviewer 2: I assume AssumRole logic should take effect automatically.a

   - **DELETE** `/v1/backupvolumes/{volName}?action=backupDelete`: delete a single volume backup from the cluster custom resource (CR) and also from the remote backup store.

### Test plan

With over 1k backup volumes and over 1k historical volume backups under pretty high network latency (700-800ms per operation)
from longhorn manager to the remote backup store:
1. The user can list backup volumes and list historical volume backups on the Longhorn GUI.
2. When the user deletes a backup volume on the Longhorn GUI:
   1. list backup volumes on the Longhorn GUI, the deleted one does not exist immediately.
   2. check the remote backup store, the backup volume will be deleted after a while.
3. When the user deletes a historical volume backup on the Longhorn GUI:
   1. list historical volume backups on the Longhorn GUI, the deleted one does not exist immediately.
   2. check the remote backup store, the historical backup will be deleted after a while.
4. When the user deletes a backup volume or deletes a historical volume backup on the remote backup store manually.

>    After `backupstore-poll-interval` seconds, list backup volumes or list historical volume backups on the Longhorn GUI, the deleted one does not exist anymore.

Reviewer 2: Before this period, just make sure that any operation against to the missing volume backup should be failed w/ error message.

   | **DELETE** `/v1/backupvolumes/{volName}?action=backupDelete` | delete a backup from the remote backup target         | delete the Backup CR, `backup_controller` reconciles to delete a backup from reomte backup target                                                    |

## Design

### Implementation Overview

1. Create a new BackupTarget CRD `backuptargets.longhorn.io`.

   ```yaml
   metadata:
     name: the backup target name (`default`), since currently we only support one backup target.
   spec:
     backupTargetURL: the backup target URL. (string)
     credentialSecret: the backup target credential secret. (string)
     pollInterval: the backup target poll interval. (metav1.Duration)
     forceSync: to force sync the backup volume or not. (bool)
   status:
     available: records if the remote backup target is available or not. (bool)
     lastSyncedAt: records the last time the backup target was running the reconcile process. (*metav1.Time)
   ```

2. Create a new BackupVolume CRD `backupvolumes.longhorn.io`.

   ```yaml
   metadata:
     name: the backup volume name.
   spec:
     forceSync: to force sync the backup volume or not. (bool)

>      configModificationTime: the backup volume config modification time. (Time)

Reviewer: 1. Do we really need this? I don't find when the controllers will use it.
2. If YES, do not put this field in `spec`.

   ```yaml
   metadata:
     name: the backup name.
     labels:
       longhornvolume=<backup-volume-name>`: this label indicates which backup volume the backup belongs to.
   spec:
     snapshotName: the snapshot name. (string)
     labels: the labels of snapshot backup. (map[string]string)
     backingImage: the backing image. (string)
     backingImageURL: the backing image URL. (string)
     forceSync: to force sync the backup volume or not. (bool)
     configModificationTime: the backup config modification time. (Time)
   status:
     backupCreate: to indicate the backup been created or not. (bool)
     url: the snapshot backup URL. (string)
     snapshotName: the snapshot name. (string)
     snapshotCreateAt: the snapshot creation time. (string)
     backupCreateAt: the snapshot backup creation time. (string)
     size: the snapshot size. (string)
     labels: the labels of snapshot backup. (map[string]string)
     messages: the error messages when calling longhorn engine on listing or inspecting backups. (map[string]string)
     lastSyncedAt: records the last time the backup was synced into the cluster. (*metav1.Time)
   ```

4. At the existed `setting_controller`.

>    Watches the changes of Setting CR `settings.longorn.io` field `backup-target`, `backup-target-credential-secret`, and `backupstore-poll-interval`. The setting controller is responsible to create/update the default BackupTarget CR.

Reviewer: What will happen if the backup target changes? Add it's better to add a test for this case.

Author: The backup target changes did nothing, the backup_target_controller reconciles to create/delete BackupVolume CRs and back_volume_controller reconciles to create/delete Backup CRs.

The only to note is when backup target is not accessible, the BackupVolume CRs and Backup CRs will be deleted.

Reviewer: If Longhorn decides to create BackupVolume CR in this HTTP API, it can set `backuptarget.spec.forceSync` to avoid waiting for the pull interval. 
If you want to keep the current design unchanged, it's better to explain this case in the doc.

Reviewer: Got it. After syncing with the new backup target, the controllers will delete all expired CRs automatically as well as creating new CRs based on the backup target contents.

5. Create a new `backup_target_controller`.

   Watches the change of BackupTarget CR. The backup target controller is responsible for creating/deleting BackupVolume CR metadata+spec. The reconcile loop steps are:
   1. Get the default BackupTarget CR.
   2. Check if the default BackupTarget CR `spec.PollInterval == 0` _or_ `spec.forceSync == false`. If yes, abort the current reconcile process. It no, continue the reconcile process.
   3. Check if `time.Now() - the default BackupTarget CR status.lastSyncedAt >= spec.pollInterval`.
     - If no, skip the current reconcile process.
     - If yes:
       1. Call the longhorn engine to list all the backup volumes `backup ls --volume-only` from the remote backup target `backupStoreBackups`. If the remote backup target is not available, updates the BackupTarget CR `status.available=false` and `status.lastSyncedAt=time.Now()`, then skip the current reconcile process.
       2. List in cluster BackupVolume CRs `clusterBackupVolumes`.

>        3. Find the difference backup volumes `backupVolumesToPull = backupStoreBackups - clusterBackupVolumes` and create BackupVolume CR `metadata.name` + `spec.forceSync = true`.

Author: Fix the typo.

       4. Find the difference backup volumes `backupVolumesToDelete = clusterBackupVolumes - backupStoreBackups` and delete BackupVolume CR.

>        5. Updates the BackupTarget CR `spec.forceSync = false` if `spec.forceSync = true`.

Author: okay!

>        5. Updates the BackupTarget CR `spec.forceSync = false` if `spec.forceSync = true`.

Author: Use `spec.forceSync` and `status.forceSynced` would be easier for the user to control directly patch the BackupTarget CR _or_ BackupVolume CR to immediately force sync the CRs.

For example:
```shell
kubectl patch lhbt default --type merge -p '{"spec":{"forceSync":true}}'

kubectl patch lhbv <backup-volume-name> --type merge -p '{"spec":{"forceSync":true}}'
```

>        5. Updates the BackupTarget CR `spec.forceSync = false` if `spec.forceSync = true`.

Author: I'm thinking that I could use the `status.LastSyncedAt` to controller whether to force sync the backup_volume_controller.

For example, in the backup_controller, after snapshot backup completion, set the BackupVolume CR `status.LastSyncedAt = nil`. At backup_volume_controller, since the `status.LastSyncedAt = nil`, it will run sync immediately.

But I'm not sure at your point, is it okay/safe for the backup_controller to update the BackupVolume CR `status` field?

>        5. Updates the BackupTarget CR `spec.forceSync = false` if `spec.forceSync = true`.

Reviewer: This is not recommended. The user should be responsible for updating the spec only. Then Kubernetes/Longhorn can do something based on the spec.
IMO, if users want to force sync a Backup or BackupVolume via CR update, they should set the timestamp correctly. we don't need to consider if the field can be manually set **easily** here.

   3. Check if BackupVolume CR `spec.forceSync == false` and if the default BackupTarget CR `spec.PollInterval == 0` _or_ `time.Now() - the default BackupTarget CR status.lastSyncedAt < spec.pollInterval`, abort the current reconcile process.
   4. Call the longhorn engine to list all the backups `backup ls --volume <volume-name>` from the remote backup target `backupStoreBackups`.
   5.  List in cluster Backup CRs `clusterBackups`.
   6.  Find the difference backups `backupsToPull = backupStoreBackups - clusterBackups` and create Backup CR `metadata.name` + `metadata.labels["longhornvolume"]=<backup-volume-name>`.
   7.  Find the difference backups `backupsToDelete = clusterBackups - backupStoreBackups` and delete Backup CR.
   8.  Call the longhorn engine to get the backup volume config's modification time `backup config-metadata <volume-config>`. If the config metadata not changed, abort the current reconcile process.
   9. Call the longhorn engine to read the backup volumes' config `backup inspect-volume <volume-name>`.

>    10. Updates the BackupVolume CR status field according to the backup volumes' config, and also updates the BackupVolume CR `status.lastSyncedAt`.

Reviewer: Should these 2 steps stand behind step 3? If there are backup updates, there must be modifications happening on the config file. 
Besides, I am thinking about the necessity of API `backup config-metadata`. If we can directly identify if there are modifications by checking the content of the config file, why do we want an extra API call? There are too much data in the config file, and checking the last modified time of the config file would save some bandwidth cost?

Author: Before backup refactoring,
- Longhorn list and read BackupVolume and/or Backup config from the remote backup store when the user clicks the `Backup` at the GUI.
- Longhorn list and read BackupVolume config within the setting_controller timer (the timer period is the poll interval).

After backup refactoring, the backup_volume_controller and backup_controller be triggered every 30 secs. It means every 30 secs, the read API be triggered. Moreover, the Read API is a time-consuming API than the HEAD API.

I was thinking that if the Backup config has been synced, do not read the Backup config again. Because in reality, the backup config should be created once.
However, when I look at the integration test, Longhorn did a test that the Backup config might be corrupted after a successful backup config was created.

Reviewer: As we discussed in the session, after the backup creation complete, the controller doesn't need to check the backup cfg file anymore. But the for the backup volume cfg file, I think the backup-volume controller would call the `head` API first then determine if we need to inspect the cfg file and list all backups when the sync is required.

   11. Updates the BackupVolume CR spec field `spec.forceSync` and `spec.configModificationTime`.
   12. Updates the Volume CR `status.lastBackup` and `status.lastBackupAt`.

8.  For the Longhorn manager HTTP endpoints:

   - **POST** `/v1/volumes/{volName}?action=snapshotBackup`:

>      1. Generate the backup name <backup-name>.

Reviewer: Need to clarify if Longhorn should create a BackupVolume CR in this API when it doesn't exist.

Author: It's a bit tricky for the volume that has never run snapshot backup before.

When the user creates volume backup, the Backup CR is created, and the BackupVolume CR is created after the BackupVolume Config be pulled from the remote backup store. (It depends on the backup store poll interval).

If the backup store poll interval is 1 hour, it means that the user could see the BackupVolume on the GUI after 1 hour. Therefore, the BackupVolume CR is created for a better user experience.

Reviewer: If Longhorn decides to create BackupVolume CR in this HTTP API, it can set `backuptarget.spec.forceSync` to avoid waiting for the pull interval. 
If you want to keep the current design unchanged, it's better to explain this case in the doc.

Author: Oops, I had a bad memory of this yesterday.

After re-read the code, Longhorn would not create BackupVolume CR in this HTTP API call.
Instead, after backup completion and the BackupVolume CR does not exist, it creates the BackupVolume CR and set `backupvolume.spec.forceSync=true` in the `backup_controller``.

But I like your approach that set `backuptarget.spec.forceSync` to avoid waiting for the pull interval :+1:

Reviewer: If Longhorn creates BackupVolume CR in the backup-controller, there is no need to set `backuptarget.spec.forceSync`.

Notice that Creating a BackupVolume CR at the beginning matches the current behavior: Before the 1st backup is created, users can find an empty BackupVolume on the UI backup page.

Author: Okay, now I implement it as when the backup completion (progress 100%) then creating the BackupVolume CR.

     2. Create a new Backup CR with
        ```yaml
        metadata:
          name: <backup-name>
          labels: 
            longhornvolume: <backup-volume-name>
        spec:
          snapshotName: <snapshot-name>
          labels: <snapshot-backup-labels>
          backingImage: <backing-image>
          backingImageURL: <backing-image-URL>
        ```
   - **DELETE** `/v1/backupvolumes/{volName}?action=backupDelete`:

>      1. Add the finalizer to the Backup CR with the given backup name.

Reviewer: Adding the finalizer here? Shouldn't be the finalizer set once the CR is created?

Author: If the Backup CR is created with the finalizer. When the user deletes the remote backup store Backup config, the backup_volume_controller will delete the Backup CR, then the backup_controller reconciles the Delete event to delete the remote backup store Backup even if has been deleted already.

To add the finalizer is when the remote backup store config also needs to do clean up. That's why I only add the finalizer when the HTTP DELETE API be called.

Reviewer: I prefer the assumption you raised here. 
When backup-controller sends a delete call to the remote backup store for a non-existing backup file, it will receive either a not-found error or a success response. Then backup-controller knows the cleanup is done and it's safe to delete the CR itself by upsetting the finalizer. With this design, the deletion workflow is consistent with other kinds of CRs.

Reviewer 3: The finalizers should be added as part of CR creation.
And only removed after validation that the necessary resources have been cleaned up.

In this case, that the backup.cfg is truly gone on the backupstore.
As @shuo-wu mentioned a missing (NotFound) backup.cfg can be treated as a successful delete.

Just to elaborate a finalizer should be used to prevent deletion, till all necessary validation has been done. 

Some examples below:
(dependent resources have been removed, volume is safe to delete, replica data has been deleted on the file system, node is gone, etc)

Author: Got it. I'll add the finalizer when the CR is created.

Author: After recap my memory. The reason _why_ I only add finalizer on the HTTP Delete API is because I want to prevent if the user deletes the BackupVolume/Backup CR by accident. The controller reconciles to delete the remote config file.
Although we could add a field `spec.toDelete`, however, I realized that adding a finalizer would be much easier and simple to understand. WDYT? @shuo-wu @joshimoo

Reviewer: When will the BackupVolume/Backup CRs be deleted by accident? During Longhorn uninstallation? For the uninstallation, Longhorn can remove/unset the backup target first then BackupVolume/Backup CR deletion won't lead to actual backup file cleanup.

Author: The users delete the BackupVolume/Backup CRs manually :(
Since the external backup is the last mile to protect the data, we should make sure the backup data won't be deleted if the user just accidentally deletes CRs.

     2. Delete a Backup CR with the given backup name.

9.  Create a new controller `backup_controller`.

    Watches the change of Backup CR. The backup controller is responsible to update the Backup CR status field and create/delete backup to/from the remote backup target. The reconcile loop steps are:
    1.  Get the default BackupTarget CR.
    2.  If the delete Backup CR event comes in and the finalizer configured, delete the backup volume from the remote backup target `backup rm --volume <volume-name> <url>`.Then, updates BackupVolume CR `spec.forceSync=true` to request backup_volume_controller to immediately reconcile the backup volume. After that, remove the finalizer.

>     3.  Check if the Backup CR `spec.snapshotName != ""` and `status.backupCreate == false`. If yes, call longhorn engine/replica for backup creation. Then fork a go routine to monitor the backup creation progress. After that, updates Backup CR `status.backupCreate = true`.

Reviewer: when will `status.backupCreateAt` be set? Can we use this field to indicate if the backup creation is complete?

Author: Yes, when the backup creation complete.

Reviewer: Then do we still need `status.backupCreate`?

Author: Yes, if the backup is in progress, the `status.BackupCreateAt` is empty. Then when the backup_controller reconciles again and sees the `status.BackupCreateAt` is empty, the back_controller calls the snapshot backup process again.

Reviewer: Oh, OK. This field actually indicates if the backup creation is **started** or not. Can you rename this field to avoid misunderstanding?

Author: Change the wording to `status.backupCreationIsStarted`.

    5.  Call the longhorn engine to get the backup config's modification time `backup config-metadata <backup-config>`. If the config metadata not changed, abort the current reconcile process.
    6.  Call the longhorn engine to read the backup config `backup inspect <backup-url>`.
    7.  Updates the Backup CR status field according to the backup config, and also updates the Backup CR `status.lastSyncedAt`.
    8.  Updates the Backup CR spec field `spec.forceSync` and `spec.configModificationTime`.

10. For the Longhorn manager HTTP endpoints:

   - **GET** `/v1/backupvolumes`: read all the BackupVolume CRs.
   - **GET** `/v1/backupvolumes/{volName}`: read a BackupVolume CR with the given volume name.
   - **GET** `/v1/backupvolumes/{volName}?action=backupList`: read a list of Backup CRs with the label filter `volume=<backup-volume-name>`.
   - **GET** `/v1/backupvolumes/{volName}?action=backupGet`: read a Backup CR with the given backup name.

### Test plan

With over 1k backup volumes and over 1k backups under pretty high network latency (700-800ms per operation)
from longhorn manager to the remote backup target:

> - Create a single cluster with 2 volumes (vol-A and vol-B).

Reviewer: It's better to mention the main purpose of each test.

   ```yaml
   metadata:
     name: the backup volume name.
   spec:
     syncRequestAt: the time to request run sync the remote backup volume. (*metav1.Time)
     deleteRemoteConfig: indicate to delete the remote backup volume config or not. (bool)
   status:
     lastModificationTime: the backup volume config last modification time. (Time)
     size: the backup volume size. (string)
     labels: the backup volume labels. (map[string]string)
     createAt: the backup volume creation time. (string)
     lastBackupName: the latest volume backup name. (string)
     lastBackupAt: the latest volume backup time. (string)
     dataStored: the backup volume block count. (string)
     messages: the error messages when call longhorn engine on list or inspect backup volumes. (map[string]string)
     lastSyncedAt: records the last time the backup volume was synced into the cluster. (*metav1.Time)
   ```

3. Create a new Backup CRD `backups.longhorn.io`.

   ```yaml
   metadata:
     name: the backup name.
     labels:
       longhornvolume=<backup-volume-name>`: this label indicates which backup volume the backup belongs to.
   spec:

>      deleteRemoteConfig: indicate to delete the remote backup config or not. (bool)

Author: backup_controller calls the longhorn/backstore to delete remote backup config (and the related block files if needed) when delete CR event comes in and `spec.deleteRemoteConfig=true`.

>      deleteRemoteConfig: indicate to delete the remote backup config or not. (bool)

Reviewer: Then the naming of this field is confusing. How about naming it to something like `fileCleanupRequired`?

>      deleteRemoteConfig: indicate to delete the remote backup config or not. (bool)

Author: Sure~

   1. Check if the current node ID == BackupTarget CR spec.responsibleNodeID. If no, skip the reconcile process.
   2. If the delete BackupVolume CR event comes in:
      1. updates Backup CRs `spec.deleteRemoteConfig=true` if BackupVolume CR `spec.deleteRemoteConfig=true`.
      2. deletes Backup CR with the given volume name.
      3. deletes the backup volume from the remote backup target `backup rm --volume <volume-name> <url>` if `spec.deleteRemoteConfig=true`.
      4. remove the finalizer.
   3. Check if the `status.lastSyncedAt < spec.syncRequestAt`. If no, skip the reconcile process.
   4. Call the longhorn engine to list all the backups `backup ls --volume <volume-name>` from the remote backup target `backupStoreBackups`.
   5.  List in cluster Backup CRs `clusterBackups`.
   6.  Find the difference backups `backupsToPull = backupStoreBackups - clusterBackups` and create Backup CR `metadata.name` + `metadata.labels["longhornvolume"]=<backup-volume-name>` + `metadata.finalizers["longhorn.io"]`.
   7.  Find the difference backups `backupsToDelete = clusterBackups - backupStoreBackups` and delete Backup CR.

>    8.  Call the longhorn engine to get the backup volume config's last modification time `backup head <volume-config>` and compares to `status.lastModificationTime`. If the config last modification time not changed, abort the current reconcile process.

Author: updated!

     2. Delete a Backup CR with the given backup name.

9.  Create a new controller `backup_controller`.

    Watches the change of Backup CR. The backup controller is responsible for updating the Backup CR status field and creating/deleting backup to/from the remote backup target. The reconcile loop steps are:
    1. Check if the current node ID == BackupTarget CR spec.responsibleNodeID. If no, skip the reconcile process.
    2. If the delete Backup CR event comes in:
        1.  delete the backup from the remote backup target `backup rm <url>` if Backup CR `spec.deleteRemoteConfig=true`.
        2.  update the BackupVolume CR `spec.syncRequestAt=time.Now()`.
        3.  remove the finalizer.
    3. Check if the Backup CR `spec.snapshotName != ""` and `status.backupCreationIsStart == false`. If yes:
        1.  call longhorn engine/replica for backup creation.
        2.  updates Backup CR `status.backupCreationIsStart = true`.
        3.  fork a go routine to monitor the backup creation progress. After backup creation finished (progress = 100):
            1.  update the BackupVolume CR `spec.syncRequestAt = time.Now()` if BackupVolume CR exist.

>             2.  create the BackupVolume CR `metadata.name` + `metadata.finalizers["longhorn.io"]` if BackupVolume CR not exist.

Reviewer: There is no need to mention the finalizers during creation. 

Similar above.

Author: updated!

   | **GET** `/v1/backupvolumes/{volName}?action=backupGet`       | read a  backup from the remote backup target          | read a Backup CR with the given backup name                                                                                                          |
   | **DELETE** `/v1/backupvolumes/{volName}?action=backupDelete` | delete a backup from the remote backup target         | delete the Backup CR, `backup_controller` reconciles to delete a backup from reomte backup target                                                    |

## Design

### Implementation Overview

1. Create a new BackupTarget CRD `backuptargets.longhorn.io`.

   ```yaml
   metadata:
     name: the backup target name (`default`), since currently we only support one backup target.
   spec:
     backupTargetURL: the backup target URL. (string)
     credentialSecret: the backup target credential secret. (string)
     pollInterval: the backup target poll interval. (metav1.Duration)
     syncRequestAt: the time to request run sync the remote backup target. (*metav1.Time)
   status:

>      ownerID: the node ID which should responsible for run sync the remote backup target. (string)

Reviewer 2: ```suggestion
     ownerID: the node ID which is responsible for running operations of backup target controller (string)
```

     available: records if the remote backup target is available or not. (bool)
     lastSyncedAt: records the last time the backup target was running the reconcile process. (*metav1.Time)
   ```

2. Create a new BackupVolume CRD `backupvolumes.longhorn.io`.

   ```yaml
   metadata:
     name: the backup volume name.
   spec:
     syncRequestAt: the time to request run sync the remote backup volume. (*metav1.Time)
     fileCleanupRequired: indicate to delete the remote backup volume config or not. (bool)
   status:

>      ownerID: the node ID which should responsible for run sync the remote backup target. (string)

Reviewer 2: ```suggestion
     ownerID: the node ID which is responsible for running operations of backup volume controller (string)
```

     size: the backup volume size. (string)
     labels: the backup volume labels. (map[string]string)
     createAt: the backup volume creation time. (string)
     lastBackupName: the latest volume backup name. (string)
     lastBackupAt: the latest volume backup time. (string)
     dataStored: the backup volume block count. (string)
     messages: the error messages when call longhorn engine on list or inspect backup volumes. (map[string]string)
     lastSyncedAt: records the last time the backup volume was synced into the cluster. (*metav1.Time)
   ```

3. Create a new Backup CRD `backups.longhorn.io`.

   ```yaml
   metadata:
     name: the backup name.
     labels:
       longhornvolume=<backup-volume-name>`: this label indicates which backup volume the backup belongs to.
   spec:
     fileCleanupRequired: indicate to delete the remote backup config or not (and the related block files if needed). (bool)
     snapshotName: the snapshot name. (string)
     labels: the labels of snapshot backup. (map[string]string)
     backingImage: the backing image. (string)
     backingImageURL: the backing image URL. (string)
   status:

>      ownerID: the node ID which should responsible for run sync the remote backup target. (string)

Reviewer 2: ```suggestion
     ownerID: the node ID which is responsible for running operations of the backup controller (string)
```
